mars/services/task/execution/ray/shuffle.py

```diff
@@ -0,0 +1,162 @@
+# Copyright 1999-2021 Alibaba Group Holding Ltd.
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#      http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import numpy as np
+from typing import List, Iterable
+
+from .....core.operand import ShuffleProxy, MapReduceOperand, OperandStage
+from .....utils import lazy_import
+from ....subtask import Subtask
+
+ray = lazy_import("ray")
+
+
+class ShuffleManager:
+    """Manage shuffle execution for ray by resolve dependencies between mappers outputs and reducers inputs based on
+    mapper and reducer index.
+    """
+
+    def __init__(self, subtask_graph):
+        self.subtask_graph = subtask_graph
+        self._proxy_subtasks = []
+        for subtask in subtask_graph:
+            chunk = subtask.chunk_graph.results[0]
+            if isinstance(chunk.op, ShuffleProxy):
+                self._proxy_subtasks.append(subtask)
+        self.num_shuffles = len(self._proxy_subtasks)
+        self.mapper_output_refs = []
+        self.mapper_indices = {}
+        self.reducer_indices = {}
+        for shuffle_index, proxy_subtask in enumerate(self._proxy_subtasks):
+            # Note that the reducers can also be mappers such as `DuplicateOperand`.
+            mapper_subtasks = subtask_graph.predecessors(proxy_subtask)
+            reducer_subtasks = subtask_graph.successors(proxy_subtask)
+            n_mappers = len(mapper_subtasks)
+            n_reducers = _get_reducer_operand(
+                reducer_subtasks[0].chunk_graph
+            ).n_reducers
+            mapper_output_arr = np.empty((n_mappers, n_reducers), dtype=object)
+            self.mapper_output_refs.append(mapper_output_arr)
+            self.mapper_indices.update(
+                {
+                    subtask: (shuffle_index, mapper_index)
+                    for mapper_index, subtask in enumerate(mapper_subtasks)
+                }
+            )
+            # reducers subtask should be sorted by reducer_index and MapReduceOperand.map should insert shuffle block
+            # in reducers order, otherwise shuffle blocks will be sent to wrong reducers.
+            sorted_filled_reducer_subtasks = self._get_sorted_filled_reducers(
+                reducer_subtasks, n_reducers
+            )
+            self.reducer_indices.update(
+                {
+                    subtask: (shuffle_index, reducer_ordinal)
+                    for reducer_ordinal, subtask in enumerate(
+                        sorted_filled_reducer_subtasks
```

Comment:
If the item of `sorted_filled_reducer_subtasks` may be None, is it should be skipped?

Reply:
As above mentioned, the item of `sorted_filled_reducer_subtasks` may be None, so it's possible to update a `{None: (shuffle_index, reducer_ordinal)}` to the `self. reducer_indices`. Is that expected?

Reply:
They are already updated to `self.reducer_indices`.